In [2]:
import pandas as pd
import numpy as np

In [3]:


data = {
    "Order_ID": [101, 102, 103, 104, 105, 106],
    "Customer": ["Amit", "Riya", "John", "Sara", "David", "Neha"],
    "Order_Date": [
        "2025-01-05",
        "2025-02-12",
        "2025-03-18",
        "2025-04-25",
        "2025-05-10",
        "2025-06-30"
    ],
    "Delivery_Date": [
        "2025-01-08",
        "2025-02-15",
        "2025-03-25",
        "2025-04-28",
        "2025-05-18",
        "2025-07-05"
    ],
    "Total": [2500, 1800, 3200, 2900, 4100, 1500],
    "Qty": [5, 3, 8, 6, 10, 2]
}

df = pd.DataFrame(data)

df

Order_ID Customer  Order_Date Delivery_Date  Total  Qty
0       101     Amit  2025-01-05    2025-01-08   2500    5
1       102     Riya  2025-02-12    2025-02-15   1800    3
2       103     John  2025-03-18    2025-03-25   3200    8
3       104     Sara  2025-04-25    2025-04-28   2900    6
4       105    David  2025-05-10    2025-05-18   4100   10
5       106     Neha  2025-06-30    2025-07-05   1500    2

In [4]:
df["Delivery_Date"] = pd.to_datetime(df["Delivery_Date"]);
df["Order_Date"] = pd.to_datetime(df["Order_Date"]);

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6 entries, 0 to 5
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   Order_ID       6 non-null      int64         
 1   Customer       6 non-null      object        
 2   Order_Date     6 non-null      datetime64[ns]
 3   Delivery_Date  6 non-null      datetime64[ns]
 4   Total          6 non-null      int64         
 5   Qty            6 non-null      int64         
dtypes: datetime64[ns](2), int64(3), object(1)
memory usage: 420.0+ bytes


In [6]:
df["Week_day"] = df["Order_Date"].dt.weekday
df["Day_Name"] = df["Order_Date"].dt.day_name()
df["Month_End"] = df["Order_Date"].dt.is_month_end
df["Delivery-Time"] = (df["Delivery_Date"])-(df["Order_Date"])
df["Delivery-Days"] = ((df["Delivery_Date"])-(df["Order_Date"])).dt.days


In [7]:
df

Order_ID Customer Order_Date Delivery_Date  Total  Qty  Week_day  \
0       101     Amit 2025-01-05    2025-01-08   2500    5         6   
1       102     Riya 2025-02-12    2025-02-15   1800    3         2   
2       103     John 2025-03-18    2025-03-25   3200    8         1   
3       104     Sara 2025-04-25    2025-04-28   2900    6         4   
4       105    David 2025-05-10    2025-05-18   4100   10         5   
5       106     Neha 2025-06-30    2025-07-05   1500    2         0   

    Day_Name  Month_End Delivery-Time  Delivery-Days  
0     Sunday      False        3 days              3  
1  Wednesday      False        3 days              3  
2    Tuesday      False        7 days              7  
3     Friday      False        3 days              3  
4   Saturday      False        8 days              8  
5     Monday       True        5 days              5

In [8]:
df["Delivery_category"] = pd.cut(
    df["Delivery-Days"],
    bins=[0,3,5,10],
    labels=["Fast","Normal","Late"]
)

In [9]:
df.head()

Order_ID Customer Order_Date Delivery_Date  Total  Qty  Week_day  \
0       101     Amit 2025-01-05    2025-01-08   2500    5         6   
1       102     Riya 2025-02-12    2025-02-15   1800    3         2   
2       103     John 2025-03-18    2025-03-25   3200    8         1   
3       104     Sara 2025-04-25    2025-04-28   2900    6         4   
4       105    David 2025-05-10    2025-05-18   4100   10         5   

    Day_Name  Month_End Delivery-Time  Delivery-Days Delivery_category  
0     Sunday      False        3 days              3              Fast  
1  Wednesday      False        3 days              3              Fast  
2    Tuesday      False        7 days              7              Late  
3     Friday      False        3 days              3              Fast  
4   Saturday      False        8 days              8              Late

In [10]:
df.set_index("Order_Date",inplace=True)

In [11]:
df.resample("YE-MAR")["Total"].sum()

Order_Date
2025-03-31    7500
2026-03-31    8500
Freq: YE-MAR, Name: Total, dtype: int64

In [18]:
df.resample("QE-MAR")["Total"].sum()

Order_Date
2025-03-31    7500
2025-06-30    8500
Freq: QE-MAR, Name: Total, dtype: int64

In [13]:
df["Total"].expanding().mean()

Order_Date
2025-01-05    2500.000000
2025-02-12    2150.000000
2025-03-18    2500.000000
2025-04-25    2600.000000
2025-05-10    2900.000000
2025-06-30    2666.666667
Name: Total, dtype: float64

In [14]:
df["Running_Avereage"] = df["Total"].rolling(window=3).mean()

In [15]:
df["Previous_Total"] = df["Total"].shift(1)

In [19]:
df["Growth"] = df["Total"].diff()

In [16]:
df["Growth%"] = (((df["Total"])-(df["Previous_Total"]))/(df["Previous_Total"])*100).round(2)

In [17]:
df

Order_ID Customer Delivery_Date  Total  Qty  Week_day   Day_Name  \
Order_Date                                                                     
2025-01-05       101     Amit    2025-01-08   2500    5         6     Sunday   
2025-02-12       102     Riya    2025-02-15   1800    3         2  Wednesday   
2025-03-18       103     John    2025-03-25   3200    8         1    Tuesday   
2025-04-25       104     Sara    2025-04-28   2900    6         4     Friday   
2025-05-10       105    David    2025-05-18   4100   10         5   Saturday   
2025-06-30       106     Neha    2025-07-05   1500    2         0     Monday   

            Month_End Delivery-Time  Delivery-Days Delivery_category  \
Order_Date                                                             
2025-01-05      False        3 days              3              Fast   
2025-02-12      False        3 days              3              Fast   
2025-03-18      False        7 days              7              Late   
2025-04-25      False        3 days              3              Fast   
2025-05-10      False        8 days              8              Late   
2025-06-30       True        5 days              5            Normal   

            Running_Avereage  Previous_Total  Growth%  
Order_Date                                             
2025-01-05               NaN             NaN      NaN  
2025-02-12               NaN          2500.0   -28.00  
2025-03-18       2500.000000          1800.0    77.78  
2025-04-25       2633.333333          3200.0    -9.38  
2025-05-10       3400.000000          2900.0    41.38  
2025-06-30       2833.333333          4100.0   -63.41

# Key Insights

1. Highest sales observed in May (4100).
2. Maximum delivery delay: 8 days.
3. Fast deliveries mostly within 3 days.
4. Sales dropped significantly in June (-63.41%).
5. Quarterly sales (Mar Quarter): 7500.